File: javascript/json-transform/src/__tests__/functions/TransformerFunctionEval.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";
import BigNumber from "bignumber.js";
import { BigDecimal } from "../../functions/common/FunctionHelpers";

describe("TransformerFunctionEval", () => {
  class Holder {
    public value: BigNumber | null;
    constructor(value: BigNumber | null) {
      this.value = value;
    }
  }

  test("inline", async () => {
    await assertTransformation(
      {
        a: 42,
        b: "$.a",
      },
      "$$eval:$.b",
      42,
    );
  });

  test("object", async () => {
    const arr = [
      new Holder(new BigDecimal(4)),
      new Holder(new BigDecimal(2)),
      new Holder(new BigDecimal("13.45")),
      new Holder(null),
    ];
    await assertTransformation(
      arr,
      {
        $$eval: {
          $$join: ["\\$", "$avg:", "\\$", "..value"],
        },
      },
      BigDecimal(4.8625),
    );

    await assertTransformation(
      arr,
      {
        $$eval: {
          $$jsonparse: {
            $$join: ["{", '"$$avg"', ":", '"$"', ', "by":"##current.value"', "}"],
          },
        },
      },
      BigDecimal(4.8625),
    );
  });
});

File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionEvalTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Collections;

public class TransformerFunctionEvalTest extends BaseTest {


    record holder(BigDecimal value) {}
    @Test
    void object() {
        var arr = new Object[] {
                new holder(new BigDecimal(4)),
                new holder( new BigDecimal(2)),
                new holder( new BigDecimal("13.45")),
                Collections.emptyMap()
        };
        assertTransformation(arr, fromJson("""
{
  "$$eval":{
  "$$join":[
  "\\\\$","$avg:",'\\\\$','..value']
  }
}
"""), fromJson("6.483333333333333"));

        assertTransformation(arr, fromJson("""
{
  "$$eval":{"$$jsonparse":{
  "$$join":[
  "{",
  "'$$avg'",
  ":","'$'",
  ", 'by':'##current.value'",
  "}"
    ]
  }
}}
"""), fromJson("4.8625"));



    }

}
